src/Avalonia.Base/Styling/Activators/AndActivator.cs

```diff
@@ -11,47 +9,45 @@ namespace Avalonia.Styling.Activators
     internal class AndActivator : StyleActivatorBase, IStyleActivatorSink
     {
         private List<IStyleActivator>? _sources;
-        private ulong _flags;
-        private ulong _mask;
 
         public int Count => _sources?.Count ?? 0;
 
         public void Add(IStyleActivator activator)
         {
+            if (IsSubscribed)
+                throw new AvaloniaInternalException("AndActivator is already subscribed.");
             _sources ??= new List<IStyleActivator>();
             _sources.Add(activator);
         }
 
-        void IStyleActivatorSink.OnNext(bool value, int tag)
+        void IStyleActivatorSink.OnNext(bool value) => ReevaluateIsActive();
+
+        protected override bool EvaluateIsActive()
         {
-            if (value)
-            {
-                _flags |= 1ul << tag;
-            }
-            else
-            {
-                _flags &= ~(1ul << tag);
-            }
+            if (_sources is null || _sources.Count == 0)
+                return true;
 
-            if (_mask != 0)
+            var count = _sources.Count;
+            var mask = (1ul << count) - 1;
+            var flags = 0UL;
+
+            for (var i = 0; i < count; ++i)
             {
-                PublishNext(_flags == _mask);
+                if (_sources[i].GetIsActive())
```

Comment:
I see there is a difference from old code, that now we evaluate is active for each child activator, even if AndActivator.EvaluateIsActive is triggered always only on single child activator changed.
I assume it's not a big problem, as all our activators aren't expensive to recalculate is-active. It might change if we implement ":has()" selector tho.

Reply:
Yeah, there was a problem with how we did it before, in that it was susceptible to event ordering, in that the event that triggers a re-evaluation could be triggered _after_ the call to get the active state. This happens because of the way values are re-evaluated now, and that activators now apply to frames instead of individual properties. I'll try to leave an example of what happens here tomorrow.